modules/product/src/ProductViewBuilder.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace Drupal\commerce_product;
+
+use Drupal\commerce_product\Entity\ProductInterface;
+use Drupal\Core\Entity\Entity\EntityViewDisplay;
+use Drupal\Core\Entity\EntityManagerInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Entity\EntityViewBuilder;
+use Drupal\Core\Entity\FieldableEntityInterface;
+use Drupal\Core\Language\LanguageManagerInterface;
+use Drupal\Core\Render\Element;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class ProductViewBuilder extends EntityViewBuilder {
+
+  /**
+   * The product field variation renderer.
+   *
+   * @var \Drupal\commerce_product\ProductVariationFieldRenderer
+   */
+  protected $productVariationFieldRenderer;
+
+  /**
+   * Constructs a new BlockViewBuilder.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type definition.
+   * @param \Drupal\Core\Entity\EntityManagerInterface $entity_manager
+   *   The entity manager service.
+   * @param \Drupal\Core\Language\LanguageManagerInterface $language_manager
+   *   The language manager.
+   * @param \Drupal\commerce_product\ProductVariationFieldRenderer $product_variation_field_renderer
+   *   The product variation field renderer.
+   */
+  public function __construct(EntityTypeInterface $entity_type, EntityManagerInterface $entity_manager, LanguageManagerInterface $language_manager, ProductVariationFieldRenderer $product_variation_field_renderer) {
+    parent::__construct($entity_type, $entity_manager, $language_manager);
+    $this->productVariationFieldRenderer = $product_variation_field_renderer;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
+    return new static(
+      $entity_type,
+      $container->get('entity.manager'),
+      $container->get('language_manager'),
+      $container->get('commerce_product.variation_field_renderer')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildMultiple(array $build_list) {
```

Comment:
Problem is that alterBuild() method is called after hook_ENTITY_TYPE_view() and hook_entity_view() are invoked.

I think it's more useful to do the addProductVariationFields() before that, because this info can be useful to be available in those hooks. On the other hands, hook_entity_view_alter() can be implemented in those cases where it's necessary to modify the variation data.

What do you think?